### PR TITLE
Pick Add blocking callback to EnvBuilder (#474)

### PR DESCRIPTION
**Pull request by hunterlxt**
Signed-off-by: Xintao <hunterlxt@live.com>

### CODE REVIEW

**Comment by hunterlxt** [comment]
Windows build error is not related to this PR

**Comment by BusyJay** [comment]
Better fix it before releasing a new version.

File: proto/Cargo.toml [hunk 1]
```diff
@@ -27,6 +27,6 @@ protobuf = "2"
 lazy_static = { version = "1.3", optional = true }
 
 [build-dependencies]
-protobuf-build = { version = "0.11", default-features = false }
+protobuf-build = { version = "=0.11.3", default-features = false }
```
Comment (BusyJay) [inline proto/Cargo.toml:30]: Is it possible to limiting the grpcio-compiler version?

Comment (hunterlxt) [reply proto/Cargo.toml:30]: Done

**Comment by hunterlxt** [comment]
@hicqu @BusyJay windows build error has been fixed.

File: proto/Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,6 @@ protobuf = "2"
 lazy_static = { version = "1.3", optional = true }
 
 [build-dependencies]
-protobuf-build = { version = "0.11", default-features = false }
-grpcio-compiler = { path = "../compiler", version = "0.5.0", default-features = false }
+protobuf-build = { version = "=0.11.3", default-features = false }
```
Comment (BusyJay) [inline proto/Cargo.toml:30]: Why still changing this?

File: proto/Cargo.toml [hunk 3]
```diff
@@ -27,6 +27,6 @@ protobuf = "2"
...
+grpcio-compiler = { version = "=0.5.0", default-features = false }
```
Comment (BusyJay) [inline proto/Cargo.toml:31]: This is wrong. It prevents from updating and testing local compiler.

Comment (hunterlxt) [reply proto/Cargo.toml:31]: protobuf-build use grpcio-compiler from crates.io and locked it. So you mean patch grpcio-compiler with local crate in root cargo.toml?

File: proto/Cargo.toml (hunk 2 above)
Comment (hunterlxt) [reply proto/Cargo.toml:30]: protobuf-build must be 0.11.3, or it will produce code with std::future

Comment (BusyJay) [reply proto/Cargo.toml:30]: Why? Isn't it constraint by L31?

File: proto/Cargo.toml (hunk 3 above)
Comment (BusyJay) [reply proto/Cargo.toml:31]: It can be a solution.

File: proto/Cargo.toml (hunk 2 above)
Comment (hunterlxt) [reply proto/Cargo.toml:30]: yes, cargo.lock showed that there are 2 grpcio-compiler and I'll patch it later.

**Comment by hunterlxt** [comment]
@BusyJay It has been patched but 0.11.3 must still be restricted because 0.11.4 uses grpcio-compiler 0.6.0. And I suggest that protobuf-build use new version num like 0.12 because 0.11.3 and 0.11.4 are very different.

**Comment by BusyJay** [comment]
Then patch configuration is useless.

**Comment by hunterlxt** [comment]
@BusyJay Why? I did see that grpcio-compiler only appeared once in Cargo.lock

**Comment by BusyJay** [comment]
I think whether there are two versions depends on whether protobuf-build is pinned instead of patch. Deleting patch should make no difference.

**Pull request merged**